File: sdk/python/adp_sdk/adapters/langgraph.py
"""LangGraph adapter for ADP v0.3.0.

Converts between ADP manifests and LangGraph StateGraph configurations.
"""

from typing import Any

from .base import AdapterBase
from .registry import AdapterRegistry
from ..adp_model import ADP


class LangGraphAdapter(AdapterBase):
    """Adapter for LangGraph framework.

    Converts ADP flow graphs to LangGraph StateGraph configurations
    and vice versa.
    """

    framework_id = "langgraph"

    def __init__(self) -> None:
        AdapterRegistry.register(self.__class__)

    def export(self, manifest: ADP) -> dict[str, Any]:
        """Export ADP manifest to LangGraph StateGraph config."""
        data = manifest.model_dump(by_alias=True, exclude_none=True)

        # Build StateGraph from flow.graph
        flow = data.get("flow", {})
        graph = flow.get("graph", {})
        nodes = graph.get("nodes", [])
        edges = graph.get("edges", [])

        # Map node kinds to LangGraph node types
        node_map: dict[str, dict[str, Any]] = {}
        for node in nodes:
            nid = node.get("id", "")
            kind = node.get("kind", "")
            node_config: dict[str, Any] = {}

            # Set node type based on ADP kind
            if kind == "llm":
                node_config["type"] = "ChatModel"
                model_ref = node.get("model_ref")
                if model_ref:
                    node_config["model"] = model_ref
            elif kind == "tool":
                node_config["type"] = "ToolNode"
                tool_ref = node.get("tool_ref")
                if tool_ref:
                    node_config["tool"] = tool_ref
            elif kind == "router":
                node_config["type"] = "Router"
                strategy = node.get("strategy")
                if strategy:
                    node_config["strategy"] = strategy
            elif kind == "retriever":
                node_config["type"] = "Retriever"
                memory_ref = node.get("memory_ref")
                if memory_ref:
                    node_config["memory"] = memory_ref
            elif kind in ("input", "output"):
                node_config["type"] = "Start" if kind == "input" else "End"
            else:
                node_config["type"] = "Node"

            # Add params
            params = node.get("params", {})
            if params:
                node_config["params"] = params

            node_map[nid] = node_config

        # Build edges
        edge_config: dict[str, list[str]] = {}
        for edge in edges:
            frm = edge.get("from", "")
            to = edge.get("to", "")
            if frm not in edge_config:
                edge_config[frm] = []
            edge_config[frm].append(to)

        # Apply adapter hints
        adapter_hints = (data.get("runtime", {}).get("adapter_hints", {})
                         .get("langgraph", {}))

        state_graph: dict[str, Any] = {
            "nodes": node_map,
            "edges": edge_config,
        }

        # Apply langgraph-specific hints
        if checkpointer := adapter_hints.get("checkpointer"):
            state_graph["checkpointer"] = checkpointer
        if memory_store := adapter_hints.get("memory_store"):
            state_graph["store"] = memory_store
        if recursion_limit := adapter_hints.get("recursion_limit"):
            state_graph["recursion_limit"] = recursion_limit
        if stream_mode := adapter_hints.get("stream_mode"):
            state_graph["stream_mode"] = stream_mode

        return state_graph

    def import_from(self, config: dict[str, Any]) -> ADP:
        """Import LangGraph StateGraph config to ADP manifest."""
        nodes = config.get("nodes", {})
        edges = config.get("edges", {})

        # Build flow graph
        flow_nodes = []
        flow_edges = []

        for node_id, node_config in nodes.items():
            kind_map = {
                "ChatModel": "llm",
                "ToolNode": "tool",
                "Router": "router",
                "Retriever": "retriever",
                "Start": "input",
                "End": "output",
            }
            kind = kind_map.get(node_config.get("type", ""), "tool")
            model_ref = node_config.get("model")
            tool_ref = node_config.get("tool")
            memory_ref = node_config.get("memory")

            node = {"id": node_id, "kind": kind}
            if model_ref:
                node["model_ref"] = model_ref
            if tool_ref:
                node["tool_ref"] = tool_ref
            if memory_ref:
                node["memory_ref"] = memory_ref

            flow_nodes.append(node)

        for frm, tos in edges.items():
            for to in tos:
                flow_edges.append({"from": frm, "to": to})

        # Extract adapter hints
        adapter_hints = {}
        if checkpointer := config.get("checkpointer"):
            adapter_hints["checkpointer"] = checkpointer
        if memory_store := config.get("store"):
            adapter_hints["memory_store"] = memory_store
        if recursion_limit := config.get("recursion_limit"):
            adapter_hints["recursion_limit"] = recursion_limit
        if stream_mode := config.get("stream_mode"):
            adapter_hints["stream_mode"] = stream_mode

        # Build minimal ADP manifest
        from ..adp_model import (
            RuntimeEntry,
            RuntimeModel,
            GraphModel,
            FlowModel,
        )

        return ADP(
            adp_version="0.3.0",
            id=config.get("id", "imported-from-langgraph"),
            name=config.get("name"),
            runtime=RuntimeModel(execution=[RuntimeEntry(
                backend="python",
                id="langgraph",
                module="langgraph.graph",
            )]),
            flow=FlowModel(
                id="imported-flow",
                graph=GraphModel(
                    nodes=flow_nodes,
                    edges=flow_edges,
                    start_nodes=[n["id"] for n in flow_nodes if n.get("kind") == "input"],
                    end_nodes=[n["id"] for n in flow_nodes if n.get("kind") == "output"],
                ),
                extensions={"langgraph": adapter_hints} if adapter_hints else None,
            ),
            extensions={"source_framework": "langgraph"},
        )

    def roundtrip_fidelity(self) -> dict[str, str]:
        return {
            **super().roundtrip_fidelity(),
            "flow.graph": "faithful",
            "loop.termination": "faithful",  # via recursion_limit
            "memory.stores": "faithful",  # via adapter_hints
        }

File: sdk/typescript/src/adapters/langgraph.ts
/**
 * LangGraph adapter for ADP v0.3.0.
 *
 * Converts between ADP manifests and LangGraph StateGraph configurations.
 */

import { AdapterBase } from "./base";
import { AdapterRegistry } from "./registry";
import { ADP } from "../adp";

export class LangGraphAdapter extends AdapterBase {
  readonly frameworkId = "langgraph";

  constructor() {
    super();
    AdapterRegistry.register(LangGraphAdapter);
  }

  export(manifest: ADP): Record<string, unknown> {
    const data = manifest as Record<string, unknown>;

    // Build StateGraph from flow.graph
    const flow = (data.flow as Record<string, unknown>) ?? {};
    const graph = (flow.graph as Record<string, unknown>) ?? {};
    const nodes = (graph.nodes as Array<Record<string, unknown>>) ?? [];
    const edges = (graph.edges as Array<Record<string, unknown>>) ?? [];

    // Map node kinds to LangGraph node types
    const nodeMap: Record<string, Record<string, unknown>> = {};
    for (const node of nodes) {
      const nid = (node.id as string) ?? "";
      const kind = (node.kind as string) ?? "";
      const nodeConfig: Record<string, unknown> = {};

      // Set node type based on ADP kind
      if (kind === "llm") {
        nodeConfig.type = "ChatModel";
        const modelRef = node.model_ref as string | undefined;
        if (modelRef) {
          nodeConfig.model = modelRef;
        }
      } else if (kind === "tool") {
        nodeConfig.type = "ToolNode";
        const toolRef = node.tool_ref as string | undefined;
        if (toolRef) {
          nodeConfig.tool = toolRef;
        }
      } else if (kind === "router") {
        nodeConfig.type = "Router";
        const strategy = node.strategy as string | undefined;
        if (strategy) {
          nodeConfig.strategy = strategy;
        }
      } else if (kind === "retriever") {
        nodeConfig.type = "Retriever";
        const memoryRef = node.memory_ref as string | undefined;
        if (memoryRef) {
          nodeConfig.memory = memoryRef;
        }
      } else if (kind === "input") {
        nodeConfig.type = "Start";
      } else if (kind === "output") {
        nodeConfig.type = "End";
      } else {
        nodeConfig.type = "Node";
      }

      // Add params
      const params = node.params as Record<string, unknown> | undefined;
      if (params) {
        nodeConfig.params = params;
      }

      nodeMap[nid] = nodeConfig;
    }

    // Build edges
    const edgeConfig: Record<string, string[]> = {};
    for (const edge of edges) {
      const frm = (edge.from as string) ?? "";
      const to = (edge.to as string) ?? "";
      if (!edgeConfig[frm]) {
        edgeConfig[frm] = [];
      }
      edgeConfig[frm].push(to);
    }

    // Apply adapter hints
    const adapterHints = (
      (data.runtime as Record<string, unknown>)?.adapter_hints as Record<string, unknown>
    )?.langgraph as Record<string, unknown> | undefined;

    const stateGraph: Record<string, unknown> = {
      nodes: nodeMap,
      edges: edgeConfig,
    };

    // Apply langgraph-specific hints
    if (adapterHints) {
      if (adapterHints.checkpointer !== undefined) {
        stateGraph.checkpointer = adapterHints.checkpointer;
      }
      if (adapterHints.memory_store !== undefined) {
        stateGraph.store = adapterHints.memory_store;
      }
      if (adapterHints.recursion_limit !== undefined) {
        stateGraph.recursion_limit = adapterHints.recursion_limit;
      }
      if (adapterHints.stream_mode !== undefined) {
        stateGraph.stream_mode = adapterHints.stream_mode;
      }
    }

    return stateGraph;
  }

  importFrom(config: Record<string, unknown>): ADP {
    const nodes = (config.nodes as Record<string, Record<string, unknown>>) ?? {};
    const edges = (config.edges as Record<string, string[]>) ?? {};

    // Build flow graph
    const flowNodes: Array<Record<string, unknown>> = [];
    const flowEdges: Array<Record<string, unknown>> = [];

    const kindMap: Record<string, string> = {
      ChatModel: "llm",
      ToolNode: "tool",
      Router: "router",
      Retriever: "retriever",
      Start: "input",
      End: "output",
    };

    for (const [nodeId, nodeConfig] of Object.entries(nodes)) {
      const type = (nodeConfig.type as string) ?? "";
      const kind = kindMap[type] ?? "tool";
      const node: Record<string, unknown> = { id: nodeId, kind };

      if (type === "ChatModel") {
        if (nodeConfig.model) {
          node.model_ref = nodeConfig.model;
        }
      } else if (type === "ToolNode") {
        if (nodeConfig.tool) {
          node.tool_ref = nodeConfig.tool;
        }
      } else if (type === "Router") {
        if (nodeConfig.strategy) {
          node.strategy = nodeConfig.strategy;
        }
      } else if (type === "Retriever") {
        if (nodeConfig.memory) {
          node.memory_ref = nodeConfig.memory;
        }
      }

      flowNodes.push(node);
    }

    for (const [frm, tos] of Object.entries(edges)) {
      for (const to of tos) {
        flowEdges.push({ from: frm, to });
      }
    }

    // Extract adapter hints
    const adapterHints: Record<string, unknown> = {};
    if (config.checkpointer !== undefined) {
      adapterHints.checkpointer = config.checkpointer;
    }
    if (config.store !== undefined) {
      adapterHints.memory_store = config.store;
    }
    if (config.recursion_limit !== undefined) {
      adapterHints.recursion_limit = config.recursion_limit;
    }
    if (config.stream_mode !== undefined) {
      adapterHints.stream_mode = config.stream_mode;
    }

    const startNodes = flowNodes.filter((n) => n.kind === "input").map((n) => n.id as string);
    const endNodes = flowNodes.filter((n) => n.kind === "output").map((n) => n.id as string);

    return {
      adp_version: "0.3.0",
      id: (config.id as string) ?? "imported-from-langgraph",
      name: config.name as string | undefined,
      runtime: {
        execution: [
          {
            backend: "python",
            id: "langgraph",
            module: "langgraph.graph",
          },
        ],
      },
      flow: {
        id: "imported-flow",
        graph: {
          nodes: flowNodes,
          edges: flowEdges,
          start_nodes: startNodes,
          end_nodes: endNodes,
        },
        extensions: adapterHints && Object.keys(adapterHints).length > 0 ? { langgraph: adapterHints } : undefined,
      },
      evaluation: { suites: [] },
      extensions: {
        source_framework: "langgraph",
      },
    } as unknown as ADP;
  }

  roundtripFidelity(): Record<string, string> {
    return {
      ...super.roundtripFidelity(),
      "flow.graph": "faithful",
      "loop.termination": "faithful", // via recursion_limit
      "memory.stores": "faithful", // via adapter_hints
    };
  }
}
